nexus/tests/integration_tests/instances.rs

```diff
@@ -1901,6 +1901,212 @@ async fn test_instance_fails_to_boot_with_disk(
+                },
+                size: ByteCount::from_gibibytes_u32(4),
+                disk_source: params::DiskSource::Blank {
+                    block_size: params::BlockSize::try_from(512).unwrap(),
+                },
+            }),
+            params::InstanceDiskAttachment::Attach(
+                params::InstanceDiskAttach {
+                    name: Name::try_from(String::from("probablydata2"))
+                        .unwrap(),
+                },
+            ),
+        ],
+        start: true,
+    };
+
+    let url_instances = format!(
+        "/organizations/{}/projects/{}/instances",
+        ORGANIZATION_NAME, PROJECT_NAME
+    );
+
+    let builder =
+        RequestBuilder::new(client, http::Method::POST, &url_instances)
+            .body(Some(&instance_params))
+            .expect_status(Some(http::StatusCode::CREATED));
+
+    let response = NexusRequest::new(builder)
+        .authn_as(AuthnMode::PrivilegedUser)
+        .execute()
+        .await
+        .expect("Expected instance creation!");
+
+    let instance = response.parsed_body::<Instance>().unwrap();
+
+    // Assert disks are created and attached
+    let url_project_disks = format!(
+        "/organizations/{}/projects/{}/disks",
+        ORGANIZATION_NAME, PROJECT_NAME,
+    );
+    let disks: Vec<Disk> = NexusRequest::iter_collection_authn(
+        client,
+        &url_project_disks,
+        "",
+        None,
+    )
+    .await
+    .expect("failed to list disks")
+    .all_items;
+    assert_eq!(disks.len(), 2);
+
+    for disk in disks {
+        assert_eq!(disk.state, DiskState::Attached(instance.identity.id));
+    }
+}
+
+#[nexus_test]
+async fn test_instance_create_attach_disks_undo(
+    cptestctx: &ControlPlaneTestContext,
+) {
+    let client = &cptestctx.external_client;
+
+    // Test pre-reqs
+    DiskTest::new(&cptestctx).await;
+    create_org_and_project(&client).await;
+    create_disk(&client, ORGANIZATION_NAME, PROJECT_NAME, "probablydata2")
+        .await;
+    let faulted_disk =
+        create_disk(&client, ORGANIZATION_NAME, PROJECT_NAME, "probablydata3")
+            .await;
+
+    let url_project_disks = format!(
+        "/organizations/{}/projects/{}/disks",
+        ORGANIZATION_NAME, PROJECT_NAME,
+    );
+
+    // fault probablydata3
+    let apictx = &cptestctx.server.apictx;
+    let nexus = &apictx.nexus;
+    assert!(nexus
+        .set_disk_as_faulted(&faulted_disk.identity.id)
+        .await
+        .unwrap());
+
+    // Assert FAULTED
+    let disks: Vec<Disk> = NexusRequest::iter_collection_authn(
+        client,
+        &url_project_disks,
+        "",
+        None,
+    )
+    .await
+    .expect("failed to list disks")
+    .all_items;
+    assert_eq!(disks.len(), 2);
+
+    for (i, disk) in disks.iter().enumerate() {
+        if i == 0 {
+            assert_eq!(disk.state, DiskState::Detached);
+        } else {
+            assert_eq!(disk.state, DiskState::Faulted);
+        }
+    }
```

Comment:
IMO this would be more readable as:

```rust
assert_eq!(disks[0].state, DiskState::Detached);
assert_eq!(disks[1].state, DiskState::Faulted);
```

If you wanted, we could also check UUIDs here, like:

```rust
assert_eq!(disks[0].state, regular_disk.id);
assert_eq!(disks[0].state, DiskState::Detached);

assert_eq!(disks[1].identity.id, faulted_disk.id);
assert_eq!(disks[1].state, DiskState::Faulted);
```

Then we could call this same function *after* the failed instance creation function, to assert that the pre-existing disks appear the same.